rtrlib/bgpsec/bgpsec.c

```diff
@@ -0,0 +1,635 @@
+#define BGPSEC_VERSION 0
+
+/**
+ * @brief A static list that contains all supported algorithm suites.
+ */
+static const uint8_t algorithm_suites[] = {
+	RTR_BGPSEC_ALGORITHM_SUITE_1
+};
+
+/*
+ * The data for digestion must be ordered exactly like this:
+ *
+ * +------------------------------------+
+ * | Target AS Number                   |
+ * +------------------------------------+----\
+ * | Signature Segment   : N-1          |     \
+ * +------------------------------------+     |
+ * | Secure_Path Segment : N            |     |
+ * +------------------------------------+     \
+ *       ...                                  >  Data from
+ * +------------------------------------+     /   N Segments
+ * | Signature Segment   : 1            |     |
+ * +------------------------------------+     |
+ * | Secure_Path Segment : 2            |     |
+ * +------------------------------------+     /
+ * | Secure_Path Segment : 1            |    /
+ * +------------------------------------+---/
+ * | Algorithm Suite Identifier         |
+ * +------------------------------------+
+ * | AFI                                |
+ * +------------------------------------+
+ * | SAFI                               |
+ * +------------------------------------+
+ * | NLRI                               |
+ * +------------------------------------+
+ *
+ * https://tools.ietf.org/html/rfc8205#section-4.2
+ */
+
+/* The arrays are passed in "AS path order", meaning the last appeded
+ * Signature Segment / Secure_Path Segment is at the first
+ * position of the array.
+ */
+
+int rtr_bgpsec_validate_as_path(const struct rtr_bgpsec *data,
+				struct spki_table *table)
+{
+	/* The AS path validation result. */
+	enum rtr_bgpsec_rtvals retval = 0;
+
+	/* The resulting hash. */
+	unsigned char *hash_result = NULL;
+
+	/* A temporare spki record */
+	struct spki_record *tmp_key = NULL;
+
+	/* A stream that holds the data that is hashed */
+	struct stream *s = NULL;
+
+	/* Total size of required space for the stream */
+	unsigned int stream_size = 0;
+
+	/* Use a temp variable in the validation loop since we don't want to
+	 * alter data->sigs.
+	 */
+	struct rtr_signature_seg *tmp_sig = NULL;
+
+	/* Temp variable that holds the signature length of the of the
+	 * next signature segment.
+	 */
+	int tmp_sig_len = 0;
+
+	/* Check, if the parameters are not NULL */
+	if (!data || !data->path || !data->sigs || !table)
+		return RTR_BGPSEC_MISSING_DATA;
+
+	/* Check, if there are as many signature segments as there are
+	 * secure path segments
+	 */
+	if (data->path_len != data->sigs_len)
+		return RTR_BGPSEC_WRONG_SEGMENT_AMOUNT;
+
+	/* Check, if the algorithm suite is supported by RTRlib. */
+	if (rtr_bgpsec_has_algorithm_suite(data->alg) ==
+			RTR_BGPSEC_ERROR) {
+		return RTR_BGPSEC_UNSUPPORTED_ALGORITHM_SUITE;
+	}
+
+	/* Check, if the AFI is usable with BGPsec */
+	if ((data->nlri->afi != 1) &&	/* LRTR_IPV4 */
+	    (data->nlri->afi != 2)) {	/* LRTR_IPV6 */
```

Comment:
Maybe use an enum or define for the afis?

Reply:
The pity is that there are already enums, namely LRTR_IPV{4,6}, but they have the values 0 and 1. BGPsec AFIs have the values 1 and 2. It might look strange to have multiple AFI definitions within the same code base, but I guess there is no pretty way around it. I will use defines.